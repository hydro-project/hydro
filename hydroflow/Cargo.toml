[package]
name = "hydroflow"
version = "0.1.0"
edition = "2021"

[features]
default = [ "async", "macros" ]
async = [ "futures" ]
macros = [ "hydroflow_macro", "hydroflow_datalog" ]

[[example]]
name = "echoserver"

[[example]]
name = "echo_serde_json"

[[example]]
name = "chat"

[[example]]
name = "graph_reachability"

[[example]]
name = "graph_unreachability"

[[example]]
name = "three_clique"

[[example]]
name = "two_pc"

[[example]]
name = "deadlock_detector"

[[example]]
name = "kvs"

[dependencies]
bincode = "1.3"
byteorder = "1.4.3"
bytes = "1.1.0"
futures = { version = "0.3", optional = true }
hydroflow_datalog = { optional = true, path = "../hydroflow_datalog" }
hydroflow_lang = { path = "../hydroflow_lang" }
hydroflow_macro = { optional = true, path = "../hydroflow_macro" }
pin-project-lite = "0.2.9"
pusherator = { path = "../pusherator" }
ref-cast = "1.0"
regex = "1"
sealed = "0.4"
serde = { version = "1", features = [ "derive" ] }
serde_json = "1"
slotmap = "1.0"
static_assertions = "1.1.0"
tokio = { version = "1.16", features = [ "full" ] }
tokio-stream = { version = "0.1.10", features = [ "io-util" ] }
tokio-util = { version = "0.7.4", features = [ "net", "codec" ] }
variadics = { path = "../variadics" }

[dev-dependencies]
chrono = { version = "0.4.20", features = [ "serde" ], default-features = false }
clap = { version = "3.0.14", features = [ "derive" ] }
colored = "2.0"
criterion = { version = "0.3", features = [ "async_tokio" ] }
futures = { version = "0.3" }
itertools = "0.10.3"
rand = "0.8.4"
textnonce = "1.0.0"
time = "0.3"
trybuild = "1.0.71"
hdrhistogram = "7"
zipf = "7.0.0"
core_affinity = "0.5.10"
