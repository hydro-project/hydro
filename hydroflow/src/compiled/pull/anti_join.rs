use rustc_hash::FxHashSet;
pub struct AntiJoin<'a, Key, V, Ipos>
where
    Key: Eq + std::hash::Hash + Clone,
    V: Eq + std::hash::Hash + Clone,
    Ipos: Iterator<Item = (Key, V)>,
{
    input_pos: Ipos,
    neg_state: &'a mut FxHashSet<Key>,
    pos_state: &'a mut FxHashSet<(Key, V)>,
    old_state: std::vec::IntoIter<(Key, V)>,
}

impl<'a, Key, V, Ipos> Iterator for AntiJoin<'a, Key, V, Ipos>
where
    Key: Eq + std::hash::Hash + Clone,
    V: Eq + std::hash::Hash + Clone,
    Ipos: Iterator<Item = (Key, V)>,
{
    type Item = (Key, V);

    fn next(&mut self) -> Option<Self::Item> {
        for item in self.old_state.by_ref() {
            if !self.neg_state.contains(&item.0) {
                return Some(item);
            }
        }

        for item in self.input_pos.by_ref() {
            if !self.neg_state.contains(&item.0) && !self.pos_state.contains(&item) {
                self.pos_state.insert(item.clone());
                return Some(item);
            }
        }

        None
    }
}

impl<'a, Key, V, Ipos> AntiJoin<'a, Key, V, Ipos>
where
    Key: Eq + std::hash::Hash + Clone,
    V: Eq + std::hash::Hash + Clone,
    Ipos: Iterator<Item = (Key, V)>,
{
    pub fn new(
        input_pos: Ipos,
        state: &'a mut (&'a mut FxHashSet<Key>, &'a mut FxHashSet<(Key, V)>),
        new_tick: bool,
    ) -> Self {
        Self::new_from_mut(input_pos, state.0, state.1, new_tick)
    }

    pub fn new_from_mut(
        input_pos: Ipos,
        state_neg: &'a mut FxHashSet<Key>,
        state_pos: &'a mut FxHashSet<(Key, V)>,
        new_tick: bool,
    ) -> Self {
        let old_state = if new_tick {
            state_pos
                .iter()
                .map(|(k, v)| (k.clone(), v.clone()))
                .collect::<Vec<_>>()
        } else {
            Default::default()
        };

        Self {
            input_pos,
            neg_state: state_neg,
            pos_state: state_pos,
            old_state: old_state.into_iter(),
        }
    }
}
