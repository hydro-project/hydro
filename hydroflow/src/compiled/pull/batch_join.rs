use std::marker::PhantomData;

use crate::lang::lattice::{LatticeRepr, Merge};

#[derive(Debug)]
pub struct BatchJoinState<L>
where
    L: LatticeRepr,
    L::Repr: Default,
{
    state: L::Repr,
}

impl<L> Default for BatchJoinState<L>
where
    L: LatticeRepr,
    L::Repr: Default,
{
    fn default() -> Self {
        Self {
            state: Default::default(),
        }
    }
}

pub struct BatchJoin<'a, Buf, Stream, L, Update, Tick>
where
    Buf: Iterator<Item = Update::Repr>,
    Stream: Iterator<Item = Tick>,
    Update: LatticeRepr,
    L: LatticeRepr + Merge<Update>,
    L::Repr: Default,
{
    buf: Buf,
    stream: Stream,
    state: &'a mut BatchJoinState<L>,
    _marker: PhantomData<(Update, Tick)>,
}

impl<'a, Buf, Stream, L, Update, Tick> Iterator for BatchJoin<'a, Buf, Stream, L, Update, Tick>
where
    Buf: Iterator<Item = Update::Repr>,
    Stream: Iterator<Item = Tick>,
    Update: LatticeRepr,
    L: LatticeRepr + Merge<Update>,
    L::Repr: Default,
{
    type Item = (Tick, L::Repr);

    fn next(&mut self) -> Option<Self::Item> {
        for p in &mut self.buf {
            <L as Merge<Update>>::merge(&mut self.state.state, p);
        }

        self.stream
            .next()
            .map(|t| (t, std::mem::take(&mut self.state.state)))
    }
}
impl<'a, Buf, Stream, L, Update, Tick> BatchJoin<'a, Buf, Stream, L, Update, Tick>
where
    Buf: Iterator<Item = Update::Repr>,
    Stream: Iterator<Item = Tick>,
    Update: LatticeRepr,
    L: LatticeRepr + Merge<Update>,
    L::Repr: Default,
{
    pub fn new(buf: Buf, stream: Stream, state: &'a mut BatchJoinState<L>) -> Self {
        Self {
            buf,
            stream,
            state,
            _marker: PhantomData,
        }
    }
}
