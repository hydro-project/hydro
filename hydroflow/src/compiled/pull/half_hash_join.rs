use std::{collections::HashMap, marker::PhantomData};

use crate::lang::lattice::{LatticeRepr, Merge};

pub struct HalfHashJoinState<K, L>
where
    L: LatticeRepr,
{
    tab: HashMap<K, L::Repr>,
}

impl<K, L> Default for HalfHashJoinState<K, L>
where
    L: LatticeRepr,
{
    fn default() -> Self {
        Self {
            tab: HashMap::new(),
        }
    }
}

pub struct HalfHashJoin<'a, K, Buf, L, Update, Stream, StreamV>
where
    K: Eq + std::hash::Hash,
    L: LatticeRepr + Merge<Update>,
    Update: LatticeRepr,
    Buf: Iterator<Item = (K, Update::Repr)>,
    Stream: Iterator<Item = (K, StreamV)>,
{
    buf: Buf,
    stream: Stream,
    state: &'a mut HalfHashJoinState<K, L>,
    _marker: PhantomData<Update>,
}

impl<'a, K, Buf, L, Update, Stream, StreamV> Iterator
    for HalfHashJoin<'a, K, Buf, L, Update, Stream, StreamV>
where
    K: Eq + std::hash::Hash,
    L: LatticeRepr + Merge<Update>,
    L::Repr: Default,
    Update: LatticeRepr,
    Buf: Iterator<Item = (K, Update::Repr)>,
    Stream: Iterator<Item = (K, StreamV)>,
{
    type Item = (K, StreamV, L::Repr);

    fn next(&mut self) -> Option<Self::Item> {
        for (k, v) in &mut self.buf {
            <L as Merge<Update>>::merge(self.state.tab.entry(k).or_default(), v);
        }

        for (k, v) in &mut self.stream {
            if let Some(vals) = self.state.tab.get(&k) {
                return Some((k, v, vals.clone()));
            }
        }
        None
    }
}
impl<'a, K, Buf, L, Update, Stream, StreamV> HalfHashJoin<'a, K, Buf, L, Update, Stream, StreamV>
where
    K: Eq + std::hash::Hash,
    Buf: Iterator<Item = (K, Update::Repr)>,
    Stream: Iterator<Item = (K, StreamV)>,
    Update: LatticeRepr,
    L: LatticeRepr + Merge<Update>,
    L::Repr: Clone,
{
    pub fn new(buf: Buf, stream: Stream, state: &'a mut HalfHashJoinState<K, L>) -> Self {
        Self {
            buf,
            stream,
            state,
            _marker: PhantomData,
        }
    }
}
