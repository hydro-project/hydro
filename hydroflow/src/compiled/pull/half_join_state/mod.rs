mod multiset;
mod set;

// pub use half_join_state_trait::HalfJoinState;
pub use multiset::HalfMultisetJoinState;
pub use set::HalfSetJoinState;

pub type SetJoinState<Key, V1, V2> = (HalfSetJoinState<Key, V1, V2>, HalfSetJoinState<Key, V2, V1>);

pub trait HalfJoinState<Key, ValBuild, ValProbe> {
    fn build(&mut self, k: Key, v: &ValBuild) -> bool;
    fn probe(&mut self, k: &Key, v: &ValProbe);
    fn pop_match(&mut self) -> Option<(Key, ValProbe, ValBuild)>;
}
