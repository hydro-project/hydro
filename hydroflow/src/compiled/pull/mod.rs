//! Pull-based operator helpers, i.e. [`Iterator`] helpers.
#![allow(missing_docs)] // TODO(mingwei)

mod cross_join;
pub use cross_join::*;

mod symmetric_hash_join;
pub use symmetric_hash_join::*;

mod symmetric_hash_join_lattice;
pub use symmetric_hash_join_lattice::*;

mod half_join_state;
pub use half_join_state::*;
