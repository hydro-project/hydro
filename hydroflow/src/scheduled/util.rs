use std::cell::RefCell;
use std::rc::Rc;

/**
 * Creates a once channel where the [SendOnce] end can set a value which the [Once] end will receive and store.
 */
pub fn once<T>() -> (SendOnce<T>, Once<T>) {
    let channel = Rc::new(RefCell::new(None));
    let sender = SendOnce {
        channel: channel.clone(),
    };
    let once = Once::Unresolved { channel };
    (sender, once)
}

/**
 * The sending half of a once channel.
 */
pub struct SendOnce<T> {
    channel: Rc<RefCell<Option<T>>>,
}
impl<T> SendOnce<T> {
    pub fn send(self, item: T) {
        let old_item = self.channel.borrow_mut().replace(item);
        assert!(old_item.is_none());
    }
}

/**
 * The receiving half of a once channel.
 */
pub enum Once<T> {
    Resolved(T),
    Unresolved { channel: Rc<RefCell<Option<T>>> },
    Errored,
}
impl<T> Once<T> {
    pub fn get(&mut self) -> &mut T {
        match self {
            Self::Resolved(item) => return item,
            Self::Errored => panic!("Initialization previously failed."),
            _ => {}
        }

        *self = match std::mem::replace(self, Self::Errored) {
            Self::Unresolved { channel } => {
                let item = Rc::try_unwrap(channel)
                    .ok()
                    .expect("Initialization failed: channel is still live.")
                    .into_inner()
                    .expect("Initialization failed: item never sent.");
                Self::Resolved(item)
            }
            v => v,
        };
        self.get()
    }
}
