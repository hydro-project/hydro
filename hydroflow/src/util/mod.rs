//! Helper utilities for the Hydroflow surface syntax.

mod udp;
pub use udp::*;
mod tcp;
pub use tcp::*;

use std::net::SocketAddr;
use std::task::{Context, Poll};

use bincode;
use futures::Stream;
use serde::{Deserialize, Serialize};

pub fn unbounded_channel<T>() -> (
    tokio::sync::mpsc::UnboundedSender<T>,
    tokio_stream::wrappers::UnboundedReceiverStream<T>,
) {
    let (send, recv) = tokio::sync::mpsc::unbounded_channel();
    let recv = tokio_stream::wrappers::UnboundedReceiverStream::new(recv);
    (send, recv)
}

/// Collects the immediately available items into a `FromIterator` collection.
///
/// This consumes the stream, use [`futures::StreamExt::by_ref()`] if you want
/// to retain ownership of your stream.
pub fn recv_into<C, S>(stream: S) -> C
where
    C: FromIterator<S::Item>,
    S: Stream,
{
    let mut stream = Box::pin(stream);
    std::iter::from_fn(|| {
        match stream
            .as_mut()
            .poll_next(&mut Context::from_waker(futures::task::noop_waker_ref()))
        {
            Poll::Ready(opt) => opt,
            Poll::Pending => None,
        }
    })
    .collect()
}

pub fn serialize_msg<T>(msg: T) -> bytes::Bytes
where
    T: Serialize + for<'a> Deserialize<'a> + Clone,
{
    bytes::Bytes::from(bincode::serialize(&msg).unwrap())
}

pub fn deserialize_simple<T>(msg: bytes::BytesMut) -> T
where
    T: Serialize + for<'a> Deserialize<'a> + Clone,
{
    bincode::deserialize(&msg).unwrap()
}

pub fn deserialize_msg<T>(
    msg: Result<(bytes::BytesMut, SocketAddr), tokio_util::codec::LinesCodecError>,
) -> T
where
    T: Serialize + for<'a> Deserialize<'a> + Clone,
{
    bincode::deserialize(&(msg.unwrap().0)).unwrap()
}

pub fn ipv4_resolve(addr: &str) -> Result<SocketAddr, std::io::Error> {
    use std::net::ToSocketAddrs;
    let mut addrs = addr.to_socket_addrs()?;
    let result = addrs.find(|addr| addr.is_ipv4());
    match result {
        Some(addr) => Ok(addr),
        None => Err(std::io::Error::new(
            std::io::ErrorKind::Other,
            "Unable to resolve IPv4 address",
        )),
    }
}

pub async fn bind_udp_bytes(addr: SocketAddr) -> (UdpSink, UdpStream) {
    let socket = tokio::net::UdpSocket::bind(addr).await.unwrap();
    udp_bytes(socket)
}

pub async fn bind_udp_lines(addr: SocketAddr) -> (UdpLinesSink, UdpLinesStream) {
    let socket = tokio::net::UdpSocket::bind(addr).await.unwrap();
    udp_lines(socket)
}
