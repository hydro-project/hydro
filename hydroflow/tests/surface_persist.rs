use hydroflow::compiled::pull::HalfMultisetJoinState;
use hydroflow::util::collect_ready;
use hydroflow::{assert_graphvis_snapshots, hydroflow_syntax};
use multiplatform_test::multiplatform_test;

#[multiplatform_test]
pub fn test_persist_basic() {
    let (result_send, mut result_recv) = hydroflow::util::unbounded_channel::<u32>();

    let mut hf = hydroflow_syntax! {
        repeat_iter([1])
            -> persist()
            -> fold(0, |a, b| (a + b))
            -> for_each(|x| result_send.send(x).unwrap());
    };
    assert_graphvis_snapshots!(hf);

    for tick in 0..10 {
        assert_eq!(tick, hf.current_tick());
        hf.run_tick();
    }
    assert_eq!(
        &[1, 3, 6, 10, 15, 21, 28, 36, 45, 55],
        &*collect_ready::<Vec<_>, _>(&mut result_recv)
    );
}

#[multiplatform_test]
pub fn test_persist_pull() {
    let (result_send, mut result_recv) = hydroflow::util::unbounded_channel::<u32>();

    let mut hf = hydroflow_syntax! {
        // Structured to ensure `persist()` is pull-based.
        repeat_iter([1]) -> m0;
        null() -> m0;
        m0 = union() -> persist() -> m1;
        null() -> m1;
        m1 = union()
            -> fold(0, |a, b| (a + b))
            -> for_each(|x| result_send.send(x).unwrap());
    };
    assert_graphvis_snapshots!(hf);

    for tick in 0..10 {
        assert_eq!(tick, hf.current_tick());
        hf.run_tick();
    }
    assert_eq!(
        &[1, 3, 6, 10, 15, 21, 28, 36, 45, 55],
        &*collect_ready::<Vec<_>, _>(&mut result_recv)
    );
}

#[multiplatform_test]
pub fn test_persist_push() {
    let (result_send, mut result_recv) = hydroflow::util::unbounded_channel::<u32>();

    let mut hf = hydroflow_syntax! {
        t0 = repeat_iter([1]) -> tee();
        t0 -> null();
        t1 = t0 -> persist() -> tee();
        t1 -> null();
        t1 -> fold(0, |a, b| (a + b)) -> for_each(|x| result_send.send(x).unwrap());
    };
    assert_graphvis_snapshots!(hf);

    for tick in 0..10 {
        assert_eq!(tick, hf.current_tick());
        hf.run_tick();
    }
    assert_eq!(
        &[1, 3, 6, 10, 15, 21, 28, 36, 45, 55],
        &*collect_ready::<Vec<_>, _>(&mut result_recv)
    );
}

#[multiplatform_test]
pub fn test_persist_join() {
    let (input_send, input_recv) = hydroflow::util::unbounded_channel::<(&str, &str)>();
    let mut flow = hydroflow::hydroflow_syntax! {
        repeat_iter([("hello", "world")]) -> [0]my_join;
        source_stream(input_recv) -> persist() -> [1]my_join;
        my_join = join::<'tick>() -> for_each(|(k, (v1, v2))| println!("({}, ({}, {}))", k, v1, v2));
    };
    input_send.send(("hello", "oakland")).unwrap();
    flow.run_tick();
    input_send.send(("hello", "san francisco")).unwrap();
    flow.run_tick();
}

#[multiplatform_test]
pub fn test_persist_replay_join() {
    let (persist_input_send, persist_input) = hydroflow::util::unbounded_channel::<u32>();
    let (other_input_send, other_input) = hydroflow::util::unbounded_channel::<u32>();
    let (result_send, mut result_recv) = hydroflow::util::unbounded_channel::<(u32, u32)>();

    let mut hf = hydroflow_syntax! {
        source_stream(persist_input)
            -> persist()
            -> fold::<'tick>(0, |a, b| (a + b))
            -> next_stratum()
            -> [0]product_node;

        source_stream(other_input) -> [1] product_node;

        product_node = cross_join::<'tick, 'tick>() -> for_each(|x| result_send.send(x).unwrap());
    };
    assert_graphvis_snapshots!(hf);

    persist_input_send.send(1).unwrap();
    other_input_send.send(2).unwrap();
    hf.run_tick();
    assert_eq!(&[(1, 2)], &*collect_ready::<Vec<_>, _>(&mut result_recv));

    persist_input_send.send(2).unwrap();
    other_input_send.send(2).unwrap();
    hf.run_tick();
    assert_eq!(&[(3, 2)], &*collect_ready::<Vec<_>, _>(&mut result_recv));

    other_input_send.send(3).unwrap();
    hf.run_tick();
    assert_eq!(&[(3, 3)], &*collect_ready::<Vec<_>, _>(&mut result_recv));
}

#[multiplatform_test]
pub fn test_persist_double_handoff() {
    let (input_send, input_recv) = hydroflow::util::unbounded_channel::<usize>();
    let (input_2_send, input_2_recv) = hydroflow::util::unbounded_channel::<usize>();
    let (output_send, mut output_recv) = hydroflow::util::unbounded_channel::<(usize, usize)>();
    let mut flow = hydroflow::hydroflow_syntax! {
        teed_first_sg = source_stream(input_2_recv) -> tee();
        teed_first_sg -> [0] joined_second_sg;
        teed_first_sg -> [1] joined_second_sg;

        source_stream(input_recv) -> persist()
            -> inspect(|x| println!("LHS {} {}:{}", x, context.current_tick(), context.current_stratum())) -> [0] cross;
        joined_second_sg = cross_join::<'tick, 'tick>() -> map(|t| t.0)
            -> inspect(|x| println!("RHS {} {}:{}", x, context.current_tick(), context.current_stratum())) -> [1] cross;
        cross = cross_join::<'tick, 'tick, HalfMultisetJoinState>() -> for_each(|x| output_send.send(x).unwrap());
    };
    println!("A {}:{}", flow.current_tick(), flow.current_stratum());

    input_send.send(0).unwrap();
    flow.run_tick();
    println!("B {}:{}", flow.current_tick(), flow.current_stratum());
    assert!(collect_ready::<Vec<_>, _>(&mut output_recv).is_empty());

    input_2_send.send(1).unwrap();
    flow.run_tick();
    println!("C {}:{}", flow.current_tick(), flow.current_stratum());
    assert_eq!(&[(0, 1)], &*collect_ready::<Vec<_>, _>(&mut output_recv));
}

#[multiplatform_test]
pub fn test_persist_single_handoff() {
    let (input_send, input_recv) = hydroflow::util::unbounded_channel::<usize>();
    let (input_2_send, input_2_recv) = hydroflow::util::unbounded_channel::<usize>();
    let (output_send, mut output_recv) = hydroflow::util::unbounded_channel::<(usize, usize)>();
    let mut flow = hydroflow::hydroflow_syntax! {
        teed_first_sg = source_stream(input_2_recv) -> tee();
        teed_first_sg [0] -> null();
        teed_first_sg [1] -> joined_second_sg;
        null() -> joined_second_sg;

        source_stream(input_recv) -> persist()
            -> inspect(|x| println!("LHS {} {}:{}", x, context.current_tick(), context.current_stratum())) -> [0] cross;
        joined_second_sg = union()
            -> inspect(|x| println!("RHS {} {}:{}", x, context.current_tick(), context.current_stratum())) -> [1] cross;
        cross = cross_join::<'tick, 'tick, HalfMultisetJoinState>() -> for_each(|x| output_send.send(x).unwrap());
    };
    println!("A {}:{}", flow.current_tick(), flow.current_stratum());

    input_send.send(0).unwrap();
    flow.run_tick();
    println!("B {}:{}", flow.current_tick(), flow.current_stratum());
    assert!(collect_ready::<Vec<_>, _>(&mut output_recv).is_empty());

    input_2_send.send(1).unwrap();
    flow.run_tick();
    println!("C {}:{}", flow.current_tick(), flow.current_stratum());
    assert_eq!(&[(0, 1)], &*collect_ready::<Vec<_>, _>(&mut output_recv));
}

#[multiplatform_test]
pub fn test_persist_single_subgraph() {
    let (input_send, input_recv) = hydroflow::util::unbounded_channel::<usize>();
    let (input_2_send, input_2_recv) = hydroflow::util::unbounded_channel::<usize>();
    let (output_send, mut output_recv) = hydroflow::util::unbounded_channel::<(usize, usize)>();
    let mut flow = hydroflow::hydroflow_syntax! {
        source_stream(input_2_recv) -> joined_second_sg;

        source_stream(input_recv) -> persist()
            -> inspect(|x| println!("LHS {} {}:{}", x, context.current_tick(), context.current_stratum())) -> [0] cross;
        joined_second_sg = inspect(|x| println!("RHS {} {}:{}", x, context.current_tick(), context.current_stratum())) -> [1] cross;
        cross = cross_join::<'tick, 'tick, HalfMultisetJoinState>() -> for_each(|x| output_send.send(x).unwrap());
    };
    println!("A {}:{}", flow.current_tick(), flow.current_stratum());

    input_send.send(0).unwrap();
    flow.run_tick();
    println!("B {}:{}", flow.current_tick(), flow.current_stratum());
    assert!(collect_ready::<Vec<_>, _>(&mut output_recv).is_empty());

    input_2_send.send(1).unwrap();
    flow.run_tick();
    println!("C {}:{}", flow.current_tick(), flow.current_stratum());
    assert_eq!(&[(0, 1)], &*collect_ready::<Vec<_>, _>(&mut output_recv));
}
