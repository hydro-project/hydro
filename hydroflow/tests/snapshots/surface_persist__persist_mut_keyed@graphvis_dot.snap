---
source: hydroflow/tests/surface_persist.rs
expression: df.meta_graph().unwrap().to_dot()
---
digraph {
    subgraph "cluster n1v1" {
        fillcolor="#dddddd"
        style=filled
        label = "sg_1v1\nstratum 0"
        n1v1 [label="(n1v1) source_iter([Persist(1, 1), Persist(2, 2), Persist(3, 3), Persist(4, 4), Delete(2)])", fontname=Monaco, shape=invhouse, style = filled, color = "#0022ff", fontcolor = "#ffffff"]
        subgraph "cluster sg_1v1_var_my_tee" {
            label="var my_tee"
            n1v1
        }
    }
    subgraph "cluster n2v1" {
        fillcolor="#dddddd"
        style=filled
        label = "sg_2v1\nstratum 1"
        n2v1 [label="(n2v1) persist_mut_keyed()", fontname=Monaco, shape=invhouse, style = filled, color = "#0022ff", fontcolor = "#ffffff"]
        n3v1 [label="(n3v1) tee()", fontname=Monaco, shape=house, style = filled, color = "#ffff00"]
        n4v1 [label="(n4v1) for_each(|(_k, v)| pull_tx.send(v).unwrap())", fontname=Monaco, shape=house, style = filled, color = "#ffff00"]
        n5v1 [label="(n5v1) flat_map(|(k, v)| {\l    if v == 3 { vec![Persist(k, v), Delete(k)] } else { vec![Persist(k, v)] }\l})\l", fontname=Monaco, shape=house, style = filled, color = "#ffff00"]
        n2v1 -> n3v1
        n3v1 -> n4v1
        n3v1 -> n5v1
        subgraph "cluster sg_2v1_var_my_tee" {
            label="var my_tee"
            n2v1
            n3v1
        }
    }
    subgraph "cluster n3v1" {
        fillcolor="#dddddd"
        style=filled
        label = "sg_3v1\nstratum 2"
        n6v1 [label="(n6v1) persist_mut_keyed()", fontname=Monaco, shape=invhouse, style = filled, color = "#0022ff", fontcolor = "#ffffff"]
        n7v1 [label="(n7v1) for_each(|(_k, v)| push_tx.send(v).unwrap())", fontname=Monaco, shape=house, style = filled, color = "#ffff00"]
        n6v1 -> n7v1
    }
    n1v1 -> n8v1
    n5v1 -> n9v1
    n8v1 [label="(n8v1) handoff", fontname=Monaco, shape=parallelogram, style = filled, color = "#ddddff"]
    n8v1 -> n2v1 [arrowhead=box, color=red]
    n9v1 [label="(n9v1) handoff", fontname=Monaco, shape=parallelogram, style = filled, color = "#ddddff"]
    n9v1 -> n6v1 [arrowhead=box, color=red]
}

